liorf:
  mode: "localization"
  
  # Topics
  # input
  pointCloudTopic: "/lslidar_point_cloud"              # Point cloud data  /points_raw
  imuTopic: "/imu_raw"                        # IMU data 
  gpsTopic: "odometry/gpsz"                   # GPS odometry topic from navsat, see module_navsat.launch file

  # output
  odomTopic: "odometry/imu"                   # 最后的定位输出未解耦imu的输出

  # Frames 
  lidarFrame: "rslidar"
  baselinkFrame: "rslidar"    
  odometryFrame: "odom"
  mapFrame: "map" 

  # GPS Settings
  useImuHeadingInitialization: false          # if using GPS data, set to "true"
  useGpsElevation: false                      # if GPS elevation is bad, set to "false"
  gpsCovThreshold: 2.0                        # m^2, threshold for using GPS data
  poseCovThreshold: 25.0                      # m^2, threshold for using GPS data
  
  # Export settings  全局地图加载路径 
  savePCD: false                  
  savePCDDirectory: "/home/s402/Desktop/lio_loc_ws/src/liorf_localization/maps/xt_cheku_lio.pcd"       
  
  # Sensor Settings
  sensor: Others                              # lidar sensor type, 'Others': self calcu or 'robosense': auto calcu
  N_SCAN: 32                                  # number of lidar channel (i.e.: 16, 32, 64, 128, Livox Horizon: 6)
  Horizon_SCAN: 1800                          # lidar horizontal resolution (Velodyne:1800, Ouster:512,1024,2048, Livox Horizon: 4000)
  downsampleRate: 2                           # default: 1. Downsample your data if too many points(line). i.e., 16 = 64 / 4, 16 = 16 / 1
  point_filter_num: 1                         # default: 3. Downsample your data if too many points(point). e.g., 16: 1, 32: 5, 64: 8
  lidarMinRange: 1.0                          # default: 1.0, minimum lidar range to be used
  lidarMaxRange: 100.0                        # default: 1000.0, maximum lidar range to be used

  # IMU Settings
  imuType: -1                                  # 0: 6-axis  1: 9-axis   -1: not use imu
  imuRate: 100.0

  # hongkong ubran
  # imuAccNoise: 8.1330537434371481e-03
  # imuGyrNoise: 7.4266825125507141e-03
  # imuAccBiasN: 1.2123362494392119e-04
  # imuGyrBiasN: 8.6572985145653080e-05

  # # MULRAN Datasets
  # imuAccNoise: 0.009939570888238808e-03
  # imuGyrNoise: 0.005636343949698187e-03
  # imuAccBiasN: 0.64356659353532566e-03
  # imuGyrBiasN: 0.35640318696367613e-03

  # kitti Datasets
  imuAccNoise: 3.9939570888238808e-03
  imuGyrNoise: 1.5636343949698187e-03
  imuAccBiasN: 6.4356659353532566e-05
  imuGyrBiasN: 3.5640318696367613e-05

  imuGravity: 9.80511
  imuRPYWeight: 0.01

  # kitti  T_lb
  extrinsicTrans: [-0.81, 0.32, -0.8]
  extrinsicRot: [1, 0, 0,
                  0, 1, 0,
                  0, 0, 1]

  # # hongkong ubran 
  # Extrinsics: T_lb (lidar -> imu)
  # extrinsicTrans: [0.0, 0.0, 0.0]
  # extrinsicRot: [1, 0, 0,
  #                 0, 1, 0,
  #                 0, 0, 1]

  # # # mulran
  # Extrinsics: T_lb (lidar -> imu)
  # extrinsicTrans: [0.0, 0.0, 0.0]
  # extrinsicRot: [-1, 0, 0,
  #                 0, -1, 0,
  #                 0, 0, 1]

  # # This parameter is set only when the 9-axis IMU is used, but it must be a high-precision IMU. e.g. MTI-680
  extrinsicRPY: [-1, 0, 0,
                  0, -1, 0,
                  0, 0, 1]

  # voxel filter paprams
  mappingSurfLeafSize: 0.4                      # default: 0.4 - outdoor, 0.2 - indoor

  # robot motion constraint (in case you are using a 2D robot)
  # z轴约束、pitch、roll约束
  z_tollerance: 1000                            # meters  米-平移限制
  rotation_tollerance: 1000                     # radians 弧度-旋转限制

  # CPU Params
  numberOfCores: 12                              # number of cores for mapping optimization
  mappingProcessInterval: 0.0                    # seconds, regulate mapping frequency

  # Surrounding map
  surroundingkeyframeAddingDistThreshold: 1.0   # meters, regulate keyframe adding threshold  两帧之间的角度
  surroundingkeyframeAddingAngleThreshold: 0.2  # radians, regulate keyframe adding threshold  两帧之间的距离
  surroundingKeyframeDensity: 2.0               # meters, downsample surrounding keyframe poses   
  surroundingKeyframeSearchRadius: 50.0         # meters, within n meters scan-to-map optimization (when loop closure disabled)
  surroundingKeyframeMapLeafSize: 0.3           # downsample local map point cloud

  # ----------------------------------ignore-------------------------------------####

  # Loop closure
  loopClosureEnableFlag: false
  loopClosureFrequency: 1.0                     # Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 50                   # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 15.0             # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 30.0           # seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 25                  # number of hostory key frames will be fused into a submap for loop closure
  loopClosureICPSurfLeafSize: 0.5               # downsample icp point cloud  
  historyKeyframeFitnessScore: 0.3              # icp threshold, the smaller the better alignment

  # Visualization
  globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius
  globalMapVisualizationPoseDensity: 10.0       # meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 1.0           # meters, global map visualization cloud density
